### ISSUE TITLE
`typed.List` lowering error when appending `None` followed by a concrete type

### ISSUE THREAD

**Issue by stuartarchibald** [numba/numba#8641]
## Reporting a bug

- [x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/main/CHANGE_LOG).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

This
```python
from numba import typed

x = typed.List()
x.append(None)
x.append(1)
```
produces:
```
numba.core.errors.NumbaNotImplementedError: Failed in nopython mode pipeline (step: native lowering)
       Cannot cast int64 to none: i64 %"arg.item"
```
when trying to do the second `.append`.

**Comment by dlee992** [comment]
Indeed, `ListType` forbids to accept `Optional` during the instructor...

https://github.com/numba/numba/blob/c91070592448ebc2243442677f5358988f5feb62/numba/core/types/containers.py#L621-L623

I think we can find a way to make this possible? I'll give it a try.

**Comment by dlee992** [comment]
Yeah, I do need this support. I will come back later with a PR, if I can figure out it.
